### ISSUE TITLE
What does hs represent here?

### ISSUE THREAD

**Issue by mukherjeesougata** [s3prl/s3prl#550]
I have used the following code for feature extraction after following [Featurizer](https://github.com/s3prl/s3prl/blob/main/s3prl/nn/upstream.py#L234) 
```
>>> import torch
 >>> from s3prl.nn import S3PRLUpstream, Featurizer
  ...
 >>> model = S3PRLUpstream("xls_r_300m")
 >>> model.eval()
  ...
 >>> with torch.no_grad():
 ...     wavs = torch.randn(2, 16000 * 2)
 ...     wavs_len = torch.LongTensor([16000 * 1, 16000 * 2])
 ...     all_hs, all_hs_len = model(wavs, wavs_len)
 ...
>>> featurizer = Featurizer(model)
>>> hs, hs_len = featurizer(all_hs, all_hs_len)
```
May I know what does hs represent here?

**Comment by leo19941227** [comment]
hidden states

**Comment by mukherjeesougata** [comment]
That means all_hs represent the hidden state features of all the 24 layers in case of wav2vec2 large model?

**Comment by leo19941227** [comment]
You are correct

**Comment by leo19941227** [comment]
Oh, all hs is a list of all layers of hidden states. hs is using weighted sum to reduce all the layers into a single final output for downstream task

**Comment by mukherjeesougata** [comment]
Ater going through #187. There the following statement has been made. But here I am extracting the features and then using those to train other models of other frameworks using a defferent toolkit, tthat means feature extraction and downstreaming task are completely two different stages which does not come under one framework, then in that case how are the weights decided? 

> The weights are initialized as an uniform distribution and then trained with each downstream task.

**Comment by mukherjeesougata** [comment]
@leo19941227 Can you kindly please clarify this?

**Comment by leo19941227** [comment]
You can just directly use a list of hs without Featurizer in another toolkit, so that you can have full control on how you want to use that list of hidden states.
If you use Featurizer, the the weights are initially uniform distribution and will be updated by gradient descent along with the downstream training loss.

**Comment by leo19941227** [comment]
This behavior is toolkit-independent as long as you are using pytorch in both toolkits. If you switch to another framework like tensorflow, then the weights will not be updated and will remain uniform distribution.
